### PR TITLE
[Seq] Add a more general-purpose [RegOp]

### PR DESCRIPTION
**Pull request by fyquah**
The RegOp serves to be the everything-but-the-kitchen sink version, as per [this discourse thread](https://llvm.discourse.group/t/rfc-extending-seq-combreg-with-enables-resets-and-clears/3793/4).

```mlir
// with a negedge reset
seq.reg %d, posedge %clk, enable %enable, syncreset negedge %reset, %resetValue : !hw.struct<foo:i32> : !hw.struct<foo:i32

// with no reset
seq.reg %d, posedge %clk, enable %enable, noreset : !hw.struct<foo:i32>
```

Some implementation details:
- `%enable` is a mandatory argument. For most part, it will just be set to vdd (or `hw.constant true`)
- `%resetValue` and `%resetTo` must either both be specified, or both not be specified.

Note that the assembly for cases when there are reset is a little funky -- namely, the type of the `$input` and `$resetValue` needs to both be specified. This is mostly because we can't get tablegen to enforce an optional operand to have a type match only when it's present. I chose to have a funkier assembly format, since it's much easier than writing a custom parser & printer. I added a verifier for the RegOp to check that this requirement is enforced.

Some minor clean up:
- I shared code with as much as possible lowering compreg => SV
- I exposed some enums used in SV td to a separate file , so that we can include it in SeqOps.td

### CODE REVIEW

File: include/circt/Dialect/Seq/Seq.td [hunk 1]
```diff
@@ -47,10 +50,31 @@ def CompRegOp : SeqOp<"compreg",
     Optional<I1>:$reset, Optional<AnyType>:$resetValue);
   let results = (outs AnyType:$data);
 
-  let printer = "return ::print(p, *this);";
+  let printer = "return ::print$cppClass(p, *this);";
```
Comment (teqdruid) [inline include/circt/Dialect/Seq/Seq.td:53]: This should not be necessary as we can (and do) rely on C++ overloading.

File: include/circt/Dialect/Seq/Seq.td [hunk 2]
```diff
@@ -47,10 +50,31 @@ def CompRegOp : SeqOp<"compreg",
...
   let parser = "return ::parse$cppClass(parser, result);";
 }
 
+def RegOp : SeqOp<"reg",
+    [AllTypesMatch<["input", "data"]>, SameVariadicOperandSize]> {
+
+  let summary = "A complete implementation of registers.";
+  let description = "See the Seq dialect rationale for a longer description";
+
+  let arguments = (ins AnyType:$input, I1:$clk,
```
Comment (teqdruid) [inline include/circt/Dialect/Seq/Seq.td:63]: nit: plz be more consistent with your line breaks.

File: include/circt/Dialect/Seq/Seq.td [hunk 3]
```diff
@@ -47,10 +50,31 @@ def CompRegOp : SeqOp<"compreg",
...
+    I1:$enable,
+    Optional<I1>:$reset, Optional<AnyType>:$resetValue,
+    EventControlAttr:$clockEdge,
+    OptionalAttr<EventControlAttr>:$resetEdge,
+    ResetTypeAttr:$resetType
+    );
+
+  let results = (outs AnyType:$data);
+
+  let assemblyFormat = "$input `,` $clockEdge $clk `,` `enable` $enable `,` $resetType ($resetEdge^ $reset `,` $resetValue `:` type($resetValue))? attr-dict `:`  type($input)";
```
Comment (teqdruid) [inline include/circt/Dialect/Seq/Seq.td:73]: Rather than making the syntax awkward, I would write a custom parser.

File: lib/Dialect/Seq/SeqPasses.cpp [hunk 4]
```diff
@@ -39,38 +40,148 @@ struct CompRegLower : public OpConversionPattern<CompRegOp> {
...
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(RegOp reg, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const final;
+};
+
+} // anonymous namespace
+
+struct RegOperandsAndAttrs {
```
Comment (teqdruid) [inline lib/Dialect/Seq/SeqPasses.cpp:60]: Could you define and use a custom op interface instead of this? Looks like that's essentially what you're doing.

**Review by teqdruid — COMMENTED** [review]
I would recommend you add an entry to https://github.com/llvm/circt/blob/main/docs/RationaleSeq.md for this more general reg op first, create a PR for just that, bring it up in an ODM, then incorporate the feedback in that document, and merge it. There were a lot of strong opinions associated with the `compreg` when I did that.

File: include/circt/Dialect/Seq/Seq.td (hunk 1 above)
Comment (lattner) [reply include/circt/Dialect/Seq/Seq.td:53]: FWIW, we aren't consistent about this, but I personally prefer the `print$cppClass` - the error case is much easier to debug and reason about, because the C++ compiler says "hey I don't have `printFoo` instead of spew out 100 `print`s that aren't the right match.

Comment (teqdruid) [reply include/circt/Dialect/Seq/Seq.td:53]: Yeah, that's a good point. I haven't seen any other place in the code where we use `printFoo`, but maybe we should move over. Consistency is good.

Comment (teqdruid) [reply include/circt/Dialect/Seq/Seq.td:53]: #1412

File: include/circt/Dialect/SV/SVEnums.td [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+def AtPosEdge: I32EnumAttrCase<"AtPosEdge", 0, "posedge">;
+/// AtNegEdge triggers on a drop from 1 to 0/X/Z, or X/Z to 0.
+def AtNegEdge: I32EnumAttrCase<"AtNegEdge", 1, "negedge">;
+/// AtEdge(v) is syntactic sugar for "AtPosEdge(v) or AtNegEdge(v)".
+def AtEdge   : I32EnumAttrCase<"AtEdge", 2, "edge">;
+
+def EventControlAttr : I32EnumAttr<"EventControl", "edge control trigger",
+                                   [AtPosEdge, AtNegEdge, AtEdge]> {
+  let cppNamespace = "circt::sv";
+}
```
Comment (darthscsi) [inline include/circt/Dialect/SV/SVEnums.td:28]: This shuffling of code could be pulled out to a separate commit.  (Not one I feel needs further review).

File: include/circt/Dialect/Seq/Seq.td [hunk 6]
```diff
@@ -47,10 +50,31 @@ def CompRegOp : SeqOp<"compreg",
     Optional<I1>:$reset, Optional<AnyType>:$resetValue);
   let results = (outs AnyType:$data);
 
-  let printer = "return ::print(p, *this);";
+  let printer = "return ::print$cppClass(p, *this);";
   let parser = "return ::parse$cppClass(parser, result);";
 }
 
+def RegOp : SeqOp<"reg",
```
Comment (darthscsi) [inline include/circt/Dialect/Seq/Seq.td:57]: It seems likely that clients will use this namespace and the sv namespace.  Thus RegOp will conflict and users would have to prefix namespaces.  While not horrible, it is annoying.  Firrtl works aroudn this by having FModule and FVector for names which conflict with other common dialects to make life a bit nicer.

Comment (darthscsi) [reply include/circt/Dialect/Seq/Seq.td:57]: So maybe RegisterOp or SeqRegOp?

File: include/circt/Dialect/Seq/Seq.td [hunk 7]
```diff
@@ -47,10 +50,31 @@ def CompRegOp : SeqOp<"compreg",
...
+    [AllTypesMatch<["input", "data"]>, SameVariadicOperandSize]> {
+
+  let summary = "A complete implementation of registers.";
```
Comment (darthscsi) [inline include/circt/Dialect/Seq/Seq.td:60]: a bold claim!

File: include/circt/Dialect/Seq/Seq.td [hunk 8]
```diff
@@ -47,10 +50,31 @@ def CompRegOp : SeqOp<"compreg",
...
+  let description = "See the Seq dialect rationale for a longer description";
+
+  let arguments = (ins AnyType:$input, I1:$clk,
+    I1:$enable,
+    Optional<I1>:$reset, Optional<AnyType>:$resetValue,
```
Comment (darthscsi) [inline include/circt/Dialect/Seq/Seq.td:65]: What's the rational for reset being optional but enable not being optional?  Always enable is enable tied to 1, never reset if reset tied to 0.

**Review by darthscsi — COMMENTED** [review]
I'd split the lowering from the op definition and do that as a follow up PR.  The op definition would benefit from having the rational doc update in the same PR.

**Comment by fyquah** [comment]
I'm writing a rationale doc in https://github.com/llvm/circt/pull/1612, and will resubmit a new PR incorporating comments from there..
